cmd/juju/application/deploy.go

```diff
@@ -185,11 +185,16 @@ func newDeployCommand() *DeployCommand {
 		if err != nil {
 			return nil, errors.Trace(err)
 		}
+		s3Putter, err := apicharms.NewS3Putter(apiRoot)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		fallbackPutter := apicharms.NewFallbackPutter(s3Putter, httpPutter)
```

Comment:
Why does the juju cli code need to know this level of detail about uploading a charm? Create an http version, a s3 version then combine them into a 3 thing? Suggest this should be pushed into the api for easier use by other clients as well. 

``` code
putter, err := apicharms.NewCharmPutter(apiRoot)
```

Reply:
Looking at this, why does the juju cli client have to know about the putters at all? Why not hide in the API as an implementation detail of uploading a local charm?

Reply:
This is a bit of an artifact from a time one of our putters could not be included in `api` because it imported aws s3 libraries. So we injected this dependency.

However, since this is no longer the case given all putters are in the api package, and the client is always configured the same way, this is probably the way to go

Reply:
Updated